### ISSUE TITLE
VNLWorldStatusMessage cannot handle multiple worlds

### ISSUE THREAD

**Issue by pschord** [psforever/PSF-LoginServer#8]
PlanetSide packet decoding is so next-gen that string decoding can occur on non-byte aligned boundaries. Scodec doesnt like this (especially since codecs are structural, not dynamic in nature) and is not able to decode a _second_ world: https://github.com/psforever/PSF-LoginServer/blob/30a94a072518c02364afe5b0da2cc1fb43cdcb64/common/src/main/scala/psforever/net/VNLWorldStatusMessage.scala#L108-L111

This reason this occurs is that the end of the world record within the VNLWorldStatusMessage is not byte aligned. On the second go around for a new world record, the bitvector is misaligned. Within the PlanetSide client there is an alignment operation performed right after the length of the following string is read. This is okay, but the VNLWorldStatusMessage isn't written for a misalignment. This leads to the behavior that the first entry decodes fine, but not the second. The only way I see to handle this is to somehow communicate to the `encodedString` codec that the input stream is misaligned. But of course, scodec is _pure functional_ and PlanetSide decoding is imperative (meaning the decoding class has access to the current position in the overall bitstream).

Scodec may need to be modified or something completely custom may have to be done for this case. I would prefer to have a general solution as this issue is definitely going to come up again in a different way.

**Comment by Fate-JH** [comment]
Is the misalignment consistent enough to be able to build sub-codecs that could handle what amounts to variations?

**Comment by pschord** [comment]
Currently our solution is to have encodedString and encodedStringAligned: https://github.com/psforever/PSF-LoginServer/blob/cb2507360e85a6c66968963368b05fa33470f4e4/common/src/main/scala/net/psforever/packet/PSPacket.scala#L145-L157

But this fails in this case cause on the first server, we need encodedString, but on the second one we need encodedStringAligned. The solution to this is either find an scodec way of handling this or, just writing a custom codec that has some state

I've been putting off writing a custom codec as I feel that this problem may arise again somewhere else.

**Comment by Fate-JH** [comment]
I recently came across something similar with the `FriendsResponse` packet where the first `String` entry had 3-padding on its length field but every subsequent entry would require 7-padding on the length field to account for a `Boolean` value that travelled with the entry and keep everything byte aligned.   I solved it by first parsing the length of the data structure (mine was a `List`), then parsing the first entry with its unique `codec`, then treating the second entry as were it the first entry of the datastructure.  `scodec` magic is used to translate the packet backwards and forwards correctly.  Check out pull request #81.

I intend to try my hand at a pattern that can handle the first entry of a structure with specific `codec` and use a default `codec` for all remaining entries.

I don't think I can write a pattern that handles the first N elements with N specific `codec`s and then handles any remaining elements with a default `codec`.  That probably would require stateful behavior.
